sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml

```diff
@@ -78,14 +96,157 @@ class HasToHex a where
+
+instance HasToHex Int where
+  toHex n = Text.implode(reverse(intToHexByteList n))
+
+instance HasToHex Party where
+  toHex p = toHex(partyToText p)
+
+-- Numerics can be thought of as a decimal string with a scale. As usage sites need to know the scale, we can
+-- simply hex encode the string representation of the underlying numeric's decimal string. This allows `parseNumeric`
+-- to be used to define `HasFromHex` instances.
+numericViaStringToHex : NumericScale n => Numeric n -> BytesHex
+numericViaStringToHex num = toHex(show num)
+
 class HasFromHex a where
   -- | Converts a hex encoded string into a typed data value.
   fromHex: BytesHex -> a
 
-instance HasToHex Text where
-  toHex = primitive @"BEEncodeHex"
-
-instance HasFromHex Text where
-  fromHex "" = ""
-  fromHex arg | Text.isHex arg = primitive @"BEDecodeHex" (Text.asciiToLower arg)
-  fromHex arg = error $ "Argument is not a hex string: " <> arg
+instance HasFromHex (Optional Text) where
+  fromHex hex | Text.isHex hex = Some(primitive @"BEDecodeHex" (Text.asciiToLower hex))
+  fromHex _ = None
+
+instance HasFromHex (Optional Int) where
+  fromHex "" = Some 0
+  fromHex hex | Text.isHex hex && Text.length hex == 2 = Some(hexByteToInt hex)
+  fromHex hex | Text.isHex hex = do
+    right <- fromHex(Text.substring ((Text.length hex) - 2) 2 hex)
+    left <- fromHex(Text.substring 0 ((Text.length hex) - 2) hex)
+    Some(left * 256 + right)
+  fromHex _ = None
+
+instance HasFromHex (Optional Party) where
+  fromHex hex | Text.isHex hex = do
+    txt <- fromHex hex
+    partyFromText txt
+  fromHex _ = None
+
+-- Numerics can be thought of as a decimal string with a scale. As usage sites need to know the scale, we can
+-- simply hex encode the string representation of the underlying numeric's decimal string. This allows `parseNumeric`
+-- to be used to define `HasFromHex` instances.
+numericViaStringFromHex: NumericScale n => BytesHex -> Optional (Numeric n)
+numericViaStringFromHex hex | Text.isHex hex = do
+  txt <- fromHex hex
+  Text.parseNumeric txt
+numericViaStringFromHex _ = None
+
+-- | Number of bytes present in a byte encoded string.
+byteCount: BytesHex -> Int
+byteCount "" = 0
+byteCount hex | Text.isHex hex = (Text.length hex) / 2
+byteCount arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+-- | HIDE
+constList: Int -> a -> [a]
+constList 0 _ = []
+constList n x = x :: (constList (n - 1) x)
+
+-- | HIDE
+byteRepeat: Int -> BytesHex -> BytesHex
+byteRepeat len byte | Text.isHex byte && byteCount byte == 1 && len >= 0 = Text.implode(constList len byte)
+byteRepeat _ arg | not (Text.isHex arg) = error $ "Expected a hex encoded string but argument was: " <> arg
+byteRepeat _ hex | Text.isHex hex && byteCount hex /= 1 = error $ "Expected a byte value but hex encoded string was: " <> hex
+byteRepeat len _ = error $ "Expected a positive length but argument was: " <> show len
```

Comment:
I found the difficult to parse. 
I would rather use standard Haskell way

I would rather put the failing case first, and use the standard `otherwise` guard for the validated case. 
```suggestion
byteRepeat: Int -> BytesHex -> BytesHex
byteRepeat len arg 
  | not (Text.isHex arg) = error $ "Expected a hex encoded string but argument was: " <> arg
  |  byteCount hex /= 1  = error $ "Expected a byte value but hex encoded string was: " <> hex
  | len < 0                      = error $ "Expected a positive length but argument was: " <> show len
  | otherwise                 = Text.implode(constList len byte)
```